sdk/trace/batch_span_processor.go

```diff
@@ -0,0 +1,175 @@
+	// queue gets full it drops the spans. Use BlockOnQueueFull to change this behavior.
+	// The default value of MaxQueueSize is 2048.
+	MaxQueueSize int
+
+	// ScheduledDelayMillis is the delay interval in milliseconds between two consecutive
+	// processing of spans.
+	// The default value of ScheduledDelayMillis is 5000 msec.
+	ScheduledDelayMillis time.Duration
+
+	// MaxExportBatchSize is the maximum number of spans to process in a single batch.
+	// If there are more than one batch worth of spans then it processes multiple batches
+	// of spans one batch after the other without any delay.
+	// The default value of MaxExportBatchSize is 512.
+	MaxExportBatchSize int
+
+	// BlockOnQueueFull blocks onEnd() and onStart() method if the queue is full
+	// AND if BlockOnQueueFull is set to true.
+	// Blocking option should be used carefully as it can severely affect the performance of an
+	// application.
+	BlockOnQueueFull bool
+}
+
+// BatchSpanProcessor implements SpanProcessor interfaces. It is used by
+// exporters to receive SpanData asynchronously.
+// Use BatchSpanProcessorOption to change the behavior of the processor.
+type BatchSpanProcessor struct {
+	exporter BatchExporter
+	o        BatchSpanProcessorOption
+
+	queue   chan *SpanData
+	dropped uint32
+
+	stopCh chan struct{}
+}
+
+var _ SpanProcessor = (*BatchSpanProcessor)(nil)
+
+// NewBatchSpanProcessor creates a new instance of BatchSpanProcessor
+// for a given exporter. It returns an error if exporter is nil.
+// The newly created BatchSpanProcessor should then be registered with sdk
+// using RegisterSpanProcessor.
+func NewBatchSpanProcessor(exporter BatchExporter, opts BatchSpanProcessorOption) (*BatchSpanProcessor, error) {
+	if exporter == nil {
+		return nil, errNilExporter
+	}
+
+	bsp := &BatchSpanProcessor{
+		exporter: exporter,
+		o:        opts,
+	}
+
+	if bsp.o.ScheduledDelayMillis.Nanoseconds() == 0 {
+		bsp.o.ScheduledDelayMillis = defaultScheduledDelayMillis
+	}
+	if bsp.o.MaxQueueSize <= 0 {
+		bsp.o.MaxQueueSize = defaultMaxQueueSize
+	}
+	if bsp.o.MaxExportBatchSize <= 0 {
+		bsp.o.MaxExportBatchSize = defaultMaxExportBatchSize
+	}
+
+	bsp.queue = make(chan *SpanData, bsp.o.MaxQueueSize)
+
+	bsp.stopCh = make(chan struct{})
+
+	//Start timer to export metrics
+	ticker := time.NewTicker(bsp.o.ScheduledDelayMillis)
+	go func(ctx context.Context) {
+		defer ticker.Stop()
+		for {
+			select {
+			case <-bsp.stopCh:
+				return
+			case <-ticker.C:
+				bsp.processQueue()
+			}
+		}
+	}(context.Background())
+
+	return bsp, nil
+}
+
+// OnStart method does nothing.
+func (bsp *BatchSpanProcessor) OnStart(sd *SpanData) {
+}
+
+// OnEnd method exports SpanData using associated exporter.
+func (bsp *BatchSpanProcessor) OnEnd(sd *SpanData) {
+	bsp.enqueue(sd)
+}
+
+// Shutdown method does nothing. There is no data to cleanup.
+func (bsp *BatchSpanProcessor) Shutdown() {
+	close(bsp.queue)
+	close(bsp.stopCh)
```

Comment:
It might be better to close the queue, set a shutting down flag, and then have processQueue close the stopCh when it has exhausted the queue. Here we can then wait for the stopCh to close, so that when `Shutdown()` returns, all spans have been exported.  Currently I _think_ (if i understand correctly) calling shutdown is liable to discard spans in its pending batch.

Reply:
make sense.